File: Core/Item.cs
using KolibSoft.Catalogue.Core.Abstractions;
using static KolibSoft.Catalogue.Core.Constants;

namespace KolibSoft.Catalogue.Core;

public class Item : IItem, IValidatable, IUpdatable<Item>
{

    public Guid Id { get; set; } = default;
    public DateTime ModifiedAt { get; set; } = default;

    public virtual bool Validate(ICollection<string>? errors = default)
    {
        var valid = true;
        if (Id == Guid.Empty)
        {
            errors?.Add(InvalidId);
            valid = false;
        }
        if (ModifiedAt > DateTime.UtcNow)
        {
            errors?.Add(InvalidModification);
            valid = false;
        }
        return valid;
    }

    public void Update(Item newState)
    {
        if (Id == Guid.Empty) Id = newState.Id;
        ModifiedAt = newState.ModifiedAt;
    }

}

File: docs/lib/item.js
import { InvalidId, InvalidModification, NoGuid } from "./constants.js";
import { UTCUtils } from "./utils/utc_utils.js";

class Item {

    /** @type {string} */
    id;

    /** @type {Date} */
    modifiedAt;

    /**
     * @param {string[]} errors 
     * @returns {boolean}
     */
    validate(errors) {
        let valid = true;
        if (this.id == NoGuid) {
            errors?.push(InvalidId);
            valid = false;
        }
        if (this.modifiedAt > new Date()) {
            errors?.push(InvalidModification);
            valid = false;
        }
        return valid;
    }

    /**
     * @param {Item} newState 
     */
    update(newState) {
        if (this.id == NoGuid) this.id = newState.id;
        this.modifiedAt = newState.modifiedAt;
    }

    /**
     * @param {{id: string, modifiedAt: Date}} json 
     */
    constructor(json) {
        this.id = json?.id ?? NoGuid;
        this.modifiedAt = new Date(json?.modifiedAt ?? new Date());
    }

}

export {
    Item
}